FWCore/Common/src/OutputProcessBlockHelper.cc

```diff
@@ -0,0 +1,231 @@
+
+    unsigned int iFile = 0;
+    unsigned int innerVectorsCurrentFile = 0;
+
+    // In ProcessBlockHelper, there is a vector which contains vectors
+    // of cache indices. Iterate over the inner vectors.
+    for (auto const& innerVectorOfCacheIndices : cacheIndices) {
+      // The inner vectors are associated with input files. Several contiguous
+      // inner vectors can be associated with the same input file. Check to
+      // see if we have crossed an input file boundary and set the file
+      // index, iFile, at the next file associated with at least
+      // one inner vector if necessary.
+      while (innerVectorsCurrentFile == cacheIndexVectorsPerFile[iFile]) {
+        // Sum cache entries for all files before the current file in
+        // ProcessBlockHelper
+        fileOffset += cacheEntriesPerFile[iFile];
+        ++iFile;
+        innerVectorsCurrentFile = 0;
+      }
+      if (innerVectorsCurrentFile == 0) {
+        // Call these when the input file changes
+        setProcessOffset(iFile, nInputProcesses, nEntries, processOffset);
+        setStoredFileInProcessOffset(iFile, nInputProcesses, nEntries, storedFileInProcessOffset);
+      }
+      ++innerVectorsCurrentFile;
+
+      assert(nInputProcesses + nAddedProcesses_ == nStoredProcesses);
+
+      // Really fill the cache indices that will be stored in the output file in this loop
+      for (unsigned int iStoredProcess = 0; iStoredProcess < nStoredProcesses; ++iStoredProcess) {
+        unsigned int storedCacheIndex = invalidCacheIndex();
+        if (iStoredProcess < nInputProcesses) {
+          unsigned int cacheIndex = innerVectorOfCacheIndices[translateFromStoredIndex_[iStoredProcess]];
+          if (cacheIndex != invalidCacheIndex()) {
+            // The offsets count in the cache sequence to the first entry in
+            // the current input file and process
+            unsigned int inputOffset = fileOffset + processOffset[iStoredProcess];
+            unsigned int storedOffset = storedProcessOffset[iStoredProcess] + storedFileInProcessOffset[iStoredProcess];
+            storedCacheIndex = cacheIndex - inputOffset + storedOffset;
+          }
+        } else {
+          // This corresponds to the current process if it has newly produced
+          // ProcessBlock products (plus possibly SubProcesses).
+          storedCacheIndex = storedProcessOffset[nInputProcesses] + iStoredProcess - nInputProcesses;
+        }
+        storedCacheIndices.push_back(storedCacheIndex);
+      }
+    }
+  }
+
+  void OutputProcessBlockHelper::setStoredProcessOffset(unsigned int nInputProcesses,
+                                                        std::vector<std::vector<unsigned int>> const& nEntries,
+                                                        std::vector<unsigned int>& storedProcessOffset) const {
+    for (unsigned int iStored = 0; iStored < nInputProcesses + 1; ++iStored) {
+      storedProcessOffset[iStored] = 0;
```

Comment:
What guarantees do you have that `storedProcessOffset.size() >= nInputProcesses`?

Reply:
I changed the loop to a ranged-for loop.

Maybe it is a little less error prone and more readable this way. Definitely the more modern style. To answer your question, the vector was initialized as follows and the size of it never changes:
```
std::vector<unsigned int> storedProcessOffset(nInputProcesses + 1, 0);
```
Originally, it seemed kind of intuitive to use an integer loop since this code is all about manipulating integer indexes and the index needs to be calculated even if it is not part of the loop. Usually, I am a big fan of ranged-for loops. I like them.